src/samplers/cpuidle/mod.rs

```diff
@@ -0,0 +1,86 @@
+// Copyright 2019 Twitter, Inc.
+// Licensed under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+pub(crate) mod statistics;
+
+use self::statistics::*;
+use crate::config::Config;
+use crate::samplers::{Common, Sampler};
+
+use failure::Error;
+use logger::*;
+use metrics::*;
+use time;
+
+use std::fs;
+
+pub struct CpuIdle<'a> {
+    common: Common<'a>,
+    cores: u64,
+}
+
+fn get_cpuidle_usage(cpu: u64, state: &Statistic) -> Result<u64, ()> {
+    let filename = format!("/sys/devices/system/cpu/cpu{}/{}/usage", cpu, state);
+    Ok(fs::read_to_string(filename)
+        .map_err(|e| debug!("could not read {} for cpu{}: {}", state, cpu, e))?
+        .trim_end() // remove trailing newline
+        .parse()
+        .map_err(|e| debug!("Could not parse cpuidle usage as integer: {}", e))?)
+}
+
+impl<'a> Sampler<'a> for CpuIdle<'a> {
+    fn new(
+        config: &'a Config,
+        recorder: &'a Recorder<AtomicU32>,
+    ) -> Result<Option<Box<Self>>, Error> {
+        if config.cpuidle().enabled() {
+            Ok(Some(Box::new(CpuIdle {
+                common: Common::new(config, recorder),
+                cores: crate::common::hardware_threads().unwrap_or(1),
+            })))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn common(&self) -> &Common<'a> {
+        &self.common
+    }
+
+    fn name(&self) -> String {
+        "cpuidle".to_string()
+    }
+
+    fn sample(&mut self) -> Result<(), ()> {
+        trace!("sample {}", self.name());
+        let time = time::precise_time_ns();
+        self.register();
+        for statistic in self.common.config().cpuidle().statistics() {
+            let mut sum: u64 = 0;
+            for core in 0..self.cores {
+                sum += get_cpuidle_usage(core, &statistic)?;
+            }
+            self.common.record_counter(&statistic, time, sum);
```

Comment:
It looks like the reported numbers are in microseconds in `/sys`. We might want to scale up to nanoseconds just for consistency with other metrics?

Reply:
Done